### PR TITLE
Feat - add a possibility to set notifications for users

### PR DESCRIPTION
**Pull request by alexVasylenko**
This MR adds the possibility to set a list of users who should get notifications.
A user can create a new notification row by clicking on the button "New Notification"
A user can choose the next notification types: All, ON_FAILURE and ON_SUCCESS
A user can remove any row
A user can type the list of recipients and separate them by a comma.

test with old value - Querybook 2022-10-12 13-15-39

### CODE REVIEW

**Comment by baumandm** [comment]
Hey @alexVasylenko, we're really interested in this feature!  Are there any outstanding issues on this PR that need to be addressed to get it out of draft?  If there's anything I can help with, I'd be happy to contribute.

**Comment by jczhong84** [comment]
> > Hey @alexVasylenko, we're really interested in this feature! Are there any outstanding issues on this PR that need to be addressed to get it out of draft? If there's anything I can help with, I'd be happy to contribute.
> 
> Hi @baumandm you don't have to do anything. I'm going to resolve the conflicts and re-check this functionality again. So I think, tomorrow I will convert this MR from Draft to Ready.

Hi @alexVasylenko, thanks for the proposal here in this draft PR. @czgu and I discussed about this the other day. We'd like the notification to be like

* user can add a separate notification section for each notify type in the schedule form
* for each notify type, it will also have a "Notify To", where users can put either multiple email addresses, or slack channels/ usernames
* make sure it's **backward compatible**

we also saw a datadoc ownership transfer case recently, while it has problems on the schedules' ownership and notifications. so we'd like also make the ownership more clear by
* only the datadoc owner can create a schedule
* datadoc ownership transfer will also update the schedule's ownership (this is already fixed)
* the scheduled datadoc will always run by the datadoc owner
* datadoc editors can not edit the schedule, like readonly users

the `kwargs` field of the `task_schedule`  will look like
```
{
  "doc_id": 3,
  "exports": [],
  "user_id": 1,
  "notifications": [
    {
      "on": 0,
      "with": "slack",
      "config": {
        "to": [
          "@jianchengzhong",
          "#some-slack-channel-name"
        ]
      }
    },
    {
      "on": 0,
      "with": "email",
      "config": {
        "to": [
          "a@pinterest.com",
          "b@pinterest.com"
        ]
      }
    }
  ]
}
```
Please let us know your thoughts and if you have time to accommodate these changes, or we'll plan time for it on our end.

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 1]
```diff
@@ -284,6 +276,126 @@ export const DataDocScheduleForm: React.FunctionComponent<
...
+                        isDisabled={!notificationRow.with}
+                        options={notifyOnOptions}
+                    />
+                </div>
+
+                <SimpleField
+                    label="Notify To"
+                    name={`${notificationFormName}.config.to`}
+                    type="input"
+                    help="You can type email/slack channel here"
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:335]: tell users they can use comma `,` to separate multiple audiences

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 2]
```diff
@@ -284,6 +276,126 @@ export const DataDocScheduleForm: React.FunctionComponent<
...
+                />
+            </div>
+            <div>
+                <IconButton icon="X" onClick={handleRemoveNotifier} />
+            </div>
+        </div>
+    );
+};
+
+const NotificationsForm: React.FC<{
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:347]: nit: naming `ScheduleNotifactionsForm`

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 3]
```diff
@@ -284,6 +276,126 @@ export const DataDocScheduleForm: React.FunctionComponent<
...
+}> = ({ name, index, notificationRow, removeRow, notifiers }) => {
+    const { setFieldValue } = useFormikContext<IScheduleFormValues>();
+    const notificationFormName = `${name}[${index}]`;
+    const handleNotifyTo = useCallback((value) => {
+        setFieldValue(
+            `${notificationFormName}.config.to`,
+            value.split(',').map((v) => v.trim())
+        );
+    }, []);
+    const handleRemoveNotifier = useCallback((index) => removeRow(index), []);
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:294]: add `removeRow` to the deps array

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 4]
```diff
@@ -191,26 +202,7 @@ export const DataDocScheduleForm: React.FunctionComponent<
                             </b>
                             , and scheduled queries will also run by them.
                         </div>
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:204]: we need to update this help text accordingly. 

@czgu you mentioned last time that you'd like to have the schedule owner to be notified by default, could you elaborate?

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 5]
```diff
@@ -284,6 +276,126 @@ export const DataDocScheduleForm: React.FunctionComponent<
...
+    );
+
+    const notifyOnOptions = useMemo(
+        () =>
+            getEnumEntries(NotifyOn).map(([key, value]) => ({
+                value,
+                label: key,
+            })),
+        []
+    );
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:308]: you can probably move this to the parent component and pass them as props, as all the notification rows will share the same options

File: querybook/server/lib/scheduled_datadoc/legacy.py [hunk 6]
```diff
@@ -8,6 +11,28 @@ def set_key_if_exists(to_dict: Dict, from_dict: Dict, key: str):
...
+        schedule_config.get("notify_with")
+        and schedule_config.get("notify_on")
+        and not schedule_config.get("notifications")
+    )
+
+
+def convert_datadoc_schedule_to_notification(schedule_config):
+    user = user_logic.get_user_by_id(schedule_config.get("user_id"))
+    value = (
+        user.email if schedule_config.get("notify_with") == "email" else user.username
```
[the remark this answers is not in the document]
Comment (alexVasylenko) [reply querybook/server/lib/scheduled_datadoc/legacy.py:25]: I don't think that teams notifier uses any params from the user, all required params are in context.

File: querybook/server/lib/scheduled_datadoc/legacy.py [hunk 7]
```diff
@@ -8,6 +11,28 @@ def set_key_if_exists(to_dict: Dict, from_dict: Dict, key: str):
...
+    return [
+        {
+            "on": schedule_config.get("notify_on"),
+            "with": schedule_config.get("notify_with"),
+            "config": {"to": [notify_to]},
+        }
+    ]
+
+
 def convert_if_legacy_datadoc_schedule(schedule_config: Dict) -> Dict:
```
Comment (jczhong84) [inline querybook/server/lib/scheduled_datadoc/legacy.py:36]: A feedback from @czgu
> can we have separate convert function from schedule format v0(single export) to v1(current notification format) and to v2(new notification format), so that we can track the schedule format change history.
we can keep the function of `convert_if_legacy_datadoc_schedule` as it will be the external interface be used by the `run_data_doc` and `get_datadoc_schedule` api, but all the above two convert function in it.

File: querybook/server/lib/scheduled_datadoc/notification.py [hunk 8]
```diff
@@ -58,18 +60,35 @@ def _get_datadoc_notification_params(
...
-    notify_user(
-        user=user,
-        template_name="datadoc_completion_notification",
-        template_params=notification_params,
-        notifier_name=notify_with,
-        session=session,
-    )
+    for recepient in notify_to:
+        user = user = generate_user(notify_with, recepient)
+        notify_user(
```
Comment (jczhong84) [inline querybook/server/lib/scheduled_datadoc/notification.py:88]: as the `recipient` (also a typo above) doesn't have to be a querybook user, e.g. slack channel, so it wont make sense to keep using `notify_user` with a `user` parameter passed in. we'd think about update this function accordingly to support like slack channels.

And eventually this `notify_user` will call the `BaseNotifier.notify`, which only accepts a user as parameter, this may also need to be updated.

also did you test the slack channel notification?  the `SlackNotifier` only supports for username today, you may also need to update it as well.

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx (hunk 4 above)
Comment (jczhong84) [reply querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:198]: let's forget about the default notification for now and remove this help text

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 9]
```diff
@@ -284,6 +270,148 @@ export const DataDocScheduleForm: React.FunctionComponent<
     );
 };
 
+function getHelpInfo(selectedWith) {
+    const helpers = {
+        slack: `You can type slack channels here and you can use comma to separate multiple channels`,
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:275]: slack username or slack channel names

also how do we differentiate a username and a slack channel name?  probably we can ask user to type prefix `@` and `#` for them separately

File: querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx [hunk 10]
```diff
@@ -284,6 +270,148 @@ export const DataDocScheduleForm: React.FunctionComponent<
...
+                        label="Notify On"
+                        name={`${notificationFormName}.on`}
+                        type="react-select"
+                        isDisabled={!notificationRow.with}
+                        options={notifyOnOptions}
+                    />
+                </div>
+
+                <SimpleField
+                    label="Notify To"
```
Comment (jczhong84) [inline querybook/webapp/components/DataDocSchedule/DataDocScheduleForm.tsx:333]: In your initial PR, you demoed that you can select a user instead of typing a username. can we make it user can either freely type a username, slack channel name or an email manually, or we can provide a dropdown auto suggestion for built-in users when typing. 

This is a nice to have feature for this PR. we may do this in a separate PR.

File: querybook/server/lib/notify/notifier/slack_notifier.py [hunk 11]
```diff
@@ -18,7 +18,8 @@ def notifier_format(self):
         return "plaintext"
 
     def notify(self, user, message):
-        to = f"@{user.username}"
+        is_old_format = user.username.startswith("@") or user.username.startswith("#")
+        to = f"{user.username}" if is_old_format else f"@{user.username}"
```
Comment (alexVasylenko) [inline querybook/server/lib/notify/notifier/slack_notifier.py:22]: @jczhong84 @czgu I have done that a user can set channel or user using '#' or '@' accordingly. But, we should think about the backward compatibility and I made it like before, just convert data. But maybe we should just add a migration that can transform this value in the database?
